pkg/node/client.go

```diff
@@ -0,0 +1,165 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+
+	"github.com/kubernetes-sigs/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	cs "github.com/kubernetes-sigs/container-object-storage-interface-api/clientset/typed/objectstorage.k8s.io/v1alpha1"
+)
+
+type NodeClient struct {
+	cosiClient *cs.ObjectstorageV1alpha1Client
+	kubeClient kubernetes.Interface
+}
+
+func NewClient() (*NodeClient, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	client := cs.NewForConfigOrDie(config)
+	kube := kubernetes.NewForConfigOrDie(config)
+	return &NodeClient{
+		cosiClient: client,
+		kubeClient: kube,
+	}, nil
+}
+
+const (
+	barNameKey      = "bar-name"
+	barNamespaceKey = "bar-namespace"
+)
+
+func parseValue(key string, ctx map[string]string) (string, error) {
+	value, ok := ctx[key]
+	if !ok {
+		return "", fmt.Errorf("required volume context key unset: %v", key)
+	}
+	klog.Infof("got value: %v", value)
+	return value, nil
+}
+
+func parseVolumeContext(volCtx map[string]string) (name, ns string, err error) {
+	klog.Info("parsing bucketAccessRequest namespace/name from volume context")
+	name, err = parseValue(barNameKey, volCtx)
+	if err != nil {
+		return
+	}
+	ns, err = parseValue(barNamespaceKey, volCtx)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (n *NodeClient) getBAR(ctx context.Context, barName, barNs string) (*v1alpha1.BucketAccessRequest, error) {
+	klog.Infof("getting bucketAccessRequest %q", fmt.Sprintf("%s/%s", barNs, barName))
+	bar, err := n.cosiClient.BucketAccessRequests(barNs).Get(ctx, barName, metav1.GetOptions{})
+	if err != nil {
+		return nil, logErr(getError("bucketAccessRequest", fmt.Sprintf("%s/%s", barNs, barName), err))
+	}
+	// TODO: need to enable validation after resolving status issue - Krish
+	//if bar == nil || !bar.Status.AccessGranted {
+	//	return nil, logErr(fmt.Errorf("BucketAccessRequest is nil or BucketAccessRequest does not grant access %q", fmt.Sprintf("%s/%s", barNs, barName)))
+	//}
+	if len(bar.Spec.BucketRequestName) == 0 {
+		return nil, logErr(fmt.Errorf("bucketAccessRequest.Spec.BucketRequestName unset"))
+	}
+	return bar, nil
+}
+
+func (n *NodeClient) getBA(ctx context.Context, baName string) (*v1alpha1.BucketAccess, error) {
+	klog.Infof("getting bucketAccess %q", fmt.Sprintf("%s", baName))
+	ba, err := n.cosiClient.BucketAccesses().Get(ctx, baName, metav1.GetOptions{})
+	if err != nil {
+		return nil, logErr(getError("bucketAccess", fmt.Sprintf("%s", baName), err))
+	}
+	// TODO: need to enable validation after resolving status issue - Krish
+	//if ba == nil || !ba.Status.AccessGranted {
+	//	return nil, logErr(fmt.Errorf("BucketAccess is nil or BucketAccess does not grant access %q", fmt.Sprintf("%s", baName)))
+	//}
+	return ba, nil
+}
+
+func (n *NodeClient) getBR(ctx context.Context, brName, brNs string) (*v1alpha1.BucketRequest, error) {
```

Comment:
are there getter funcs for these basic cosi resources? I think there should be and you can call them. That we if we add clever retry logic or something to the GETs this code will get the improvement automatically.

Reply:
I am using the underlying generated API client library in these getter functions, I just layer on some additional validation that's not provided by the client. Is there a better pattern for doing this?

Reply:
That is fine and common. See if @wlan0 wants getters for these calls.